[BugFix + Refactoring] Input validation logic and display

JayChouteau opened the pull request:
Solves the bug (#301 ) for the display of the input validation for all input forms in the application. Namely, for the 3 following locations:
- Add asset
- Add asset type
- Asset type information

Furthermore, on top of the bug fixing, some refactoring was done on the logic of the input validation:
- Extracted class: all form input validation and helpers were extracted into a class called FormInputValidation
- Signature change: multiple methods had their signature changed due to the nature of the extraction as well as refactoring on the logic of the inpu validation
- Complexity and duplication reduced: The complexity and duplication were greatly reduced thanks to the 2 previous actions mentioned
- More Consistency and automation: The logic was made more automized and less hard-coded, which will allow for better maintainability. The logic was also made more consistent between the different usages of the input validation.
- Small modification to the calling of the text constants on the code that was touched.
- Comments and description updated.

Please refer to the issue to try out the steps to reproduce the bug for all input validation locations in the app.

JayChouteau committed:
REFACTORING + BugFix (#301)
- Extracted Class FormInputValidation
- Extracted methods for input validation from asset type info controller and add asset type controller to new class FormInputValidation
- Extracted methods from UIUtilities to FormInputValidation for input validation
- Automized the input validation process for asset types
- Attached the input validation to the same anchor pane as the text input controls
- Added the styling in CSS files
- Fixed the layout and logic of form input validation

JayChouteau committed:
REFACTOR and BUGFIX - Add Asset (301#)
- Signature change for the FormInputValidation methods (both asset and asset type)
- Method extraction from AddAssetController to FormInputValidation for the FormInputValidation method.
- Method extracted from inline code of the form input validation methods
- Add asset input validation fixed
- Add asset input validation automized and complexity reduced

JayChouteau committed:
Resolving merge conflicts (301#)

TalalBazerbachi reviewed (commented):
when displaying one line of input validation, I think the it should be a bit lower.
<img width="1099" alt="image" src="https://user-images.githubusercontent.com/55160650/112674902-29e01f00-8e6f-11eb-8dbf-68b171ea3ab9.png">

JayChouteau committed:
Review comments addressed: (#301)
- The error messages have been vertically aligned with their corresponding textfield
- Separation of error messages for add asset to give user more precise feedback

JayChouteau committed:
Removed unused Text Constant (#301)

JayChouteau commented:
> when displaying one line of input validation, I think the it should be a bit lower.
> <img alt="image" width="1099" src="https://user-images.githubusercontent.com/55160650/112674902-29e01f00-8e6f-11eb-8dbf-68b171ea3ab9.png">

I have aligned vertically the error messages and have made the messages more specific to exactly which error is thrown off by the input validation instead of having and/or

![verticalAlign_2](https://user-images.githubusercontent.com/47158909/112711011-b30d4b00-8e9b-11eb-9e13-8215789ea81c.PNG)
![verticalAlign](https://user-images.githubusercontent.com/47158909/112711012-b4d70e80-8e9b-11eb-9708-ef9830366b53.PNG)

najimgh commented on JavaFXUserInterface/src/main/java/utilities/FormInputValidation.java:
```diff
@@ -0,0 +1,300 @@
+     * @param nextThreshold     The Threshold succeeding
+     * @author Najim
+     */
+    private static boolean compareThresholds(TextField previousThreshold, TextField nextThreshold) {
```
Small warning from Intellij suggesting to invert the method (considered refactoring):

![Screen Shot 2021-03-27 at 2 54 21 PM](https://user-images.githubusercontent.com/46761237/112731250-77f73000-8f0c-11eb-9143-b040232f3837.png)

JayChouteau replied:
I also saw this and thought about it, but the refactoring of that method was really out of scope, and tbh, at that point, I felt like I did enough refactoring for the in put validation. Because , for this specific method, I simply moved it from where it was to this new class, but I didnt touch it. For me, it works fine. I felt like the effort was enough for this pull request

najimgh replied:
Fair enough, it is a small refactorization that doesn't affect much.

najimgh reviewed (changes requested):
Really nice job on addressing the alignment issue of the error messages. Using a field's original positioning and adding a gap to display the error message made it so there are practically no hard-coded values.
Also, very good job refactoring the input validations into a single utility class. This will provide much more flexibility for future changes. I have left a comment for one further refactoring.

Overall, the bug issue was nicely addressed and you went beyond what was asked by refactoring.

JayChouteau committed:
Merge Develop (#306)

JayChouteau committed:
Merge develop (#301)
